File: examples/go/03_expert_gateway/main.go
package main

import (
	"fmt"

	"github.com/qdrant/qql/gqql"
)

type userCtx struct {
	Tenant string
	Role   string
}

var users = map[string]userCtx{
	"alice":   {"acme", "admin"},
	"bob":     {"acme", "viewer"},
	"charlie": {"globex", "viewer"},
}

func enforce(user, query string) (string, error) {
	ctx, ok := users[user]
	if !ok {
		return "", fmt.Errorf("unknown user: %s", user)
	}
	if !gqql.IsValid(query) {
		return "", fmt.Errorf("invalid QQL query")
	}
	return gqql.InjectFilter(query, "tenant_id", "=", fmt.Sprintf(`{"str": "%s"}`, ctx.Tenant))
}

func main() {
	requests := []struct{ user, query string }{
		{"alice", "QUERY 'sales data' FROM analytics LIMIT 10"},
		{"bob", "QUERY 'sales data' FROM analytics LIMIT 10"},
		{"charlie", "QUERY 'engineering docs' FROM docs LIMIT 5"},
	}

	fmt.Println("=== QQL Query Gateway ===")
	for _, r := range requests {
		safe, err := enforce(r.user, r.query)
		if err != nil {
			fmt.Printf("  ERROR: %v\n", err)
			continue
		}
		fmt.Printf("\n  user=%-8s role=%-7s\n", r.user, users[r.user].Role)
		fmt.Printf("  raw:  %s\n", r.query)
		fmt.Printf("  safe: %s\n", trunc(safe, 130))
	}
	fmt.Println("\n  → QQL inject_filter enables auth middleware that rewrites")
	fmt.Println("    queries before they reach Qdrant — no per-tenant collections.")
}

func trunc(s string, n int) string {
	if len(s) > n {
		return s[:n] + "..."
	}
	return s
}

File: examples/rust/03_expert_gateway/src/main.rs
use std::borrow::Cow;
use qql_core::ast::{self, Value};
use qql_core::parser::Parser;

struct User { tenant: &'static str, role: &'static str }

fn enforce(user: &str, query: &str) -> String {
    let users = std::collections::HashMap::from([
        ("alice", User { tenant: "acme", role: "admin" }),
        ("bob",   User { tenant: "acme", role: "viewer" }),
        ("charlie", User { tenant: "globex", role: "viewer" }),
    ]);
    let ctx = users.get(user).unwrap();
    let mut stmt = Parser::parse(query).unwrap();
    ast::inject_filter(&mut stmt, "tenant_id", "=", &Value::Str(Cow::Owned(ctx.tenant.to_string())));
    if ctx.role == "viewer" {
        ast::inject_filter(&mut stmt, "status", "!=", &Value::Str(Cow::Borrowed("confidential")));
    }
    format!("{:#?}", stmt)
}

fn main() {
    let requests = [
        ("alice", "QUERY 'sales data' FROM analytics LIMIT 10"),
        ("bob", "QUERY 'sales data' FROM analytics LIMIT 10"),
        ("charlie", "QUERY 'engineering docs' FROM docs LIMIT 5"),
    ];
    println!("=== QQL Query Gateway ===");
    for (user, raw) in &requests {
        let safe = enforce(user, raw);
        println!("\n  raw:  {}", raw);
        println!("  safe: {}...", &safe[..safe.len().min(130)]);
    }
}
